File: app_manager/appmanager/src/main/java/app/editors/manager/ui/activities/base/BaseAppActivity.java
package app.editors.manager.ui.activities.base;

import android.os.Bundle;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import app.editors.manager.R;
import app.editors.manager.ui.interfaces.ContextDialogInterface;
import lib.toolkit.base.managers.utils.ActivitiesUtils;
import lib.toolkit.base.managers.utils.FragmentUtils;
import lib.toolkit.base.managers.utils.UiUtils;
import lib.toolkit.base.ui.activities.base.BaseActivity;

public abstract class BaseAppActivity extends BaseActivity implements FragmentManager.OnBackStackChangedListener, ContextDialogInterface {

    private final String TAG = getClass().getSimpleName();
    private static final String TAG_FINISH = "TAG_FINISH";
    private static final int TIMER_FINISH = 5000;

    protected Handler mHandler;
    protected Runnable mFinishRunnable;
    protected boolean mIsFinish;
    protected boolean mIsBackStackNotice;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        init(savedInstanceState);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(TAG_FINISH, mIsFinish);
    }

    @Override
    public void onContextDialogOpen() {

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mHandler.removeCallbacks(mFinishRunnable);
        getSupportFragmentManager().removeOnBackStackChangedListener(this);
    }

    @Override
    public void onBackStackChanged() {

    }

    private void init(final Bundle savedInstanceState) {
        initHandlers();
        initViews();
    }

    private void initHandlers() {
        mIsBackStackNotice = false;
        mHandler = new Handler();
        mFinishRunnable = () -> mIsFinish = false;
    }

    private void startResetFinishTimer() {
        mHandler.postDelayed(mFinishRunnable, TIMER_FINISH);
    }

    private void initViews() {
        getSupportFragmentManager().addOnBackStackChangedListener(this);
        mIsFinish = false;
    }

    /*
     * Fragment operations
     * */
    protected void showFragment(@NonNull final Fragment fragment, @Nullable final String tag) {
        FragmentUtils.showFragment(getSupportFragmentManager(), fragment, R.id.frame_container, tag, false);
    }

    protected void showSingleFragment(@NonNull final Fragment fragment, @Nullable final String backStackTag) {
        FragmentUtils.showSingleFragment(getSupportFragmentManager(), fragment, R.id.frame_container, backStackTag);
    }

    /*
     * Helper methods
     * */
    protected void showUrlInBrowser(final String url) {
        ActivitiesUtils.showBrowser(this, getString(R.string.chooser_web_browser), url);
    }

    protected void showEmailClients(final String to, final String subject, final String body) {
        ActivitiesUtils.showEmail(this, getString(R.string.chooser_email_client), to, subject, body);
    }

    protected void showEmailClientTemplate() {
        showEmailClients(getString(R.string.app_support_email), getString(R.string.about_email_subject),
                UiUtils.getDeviceInfoString(null, true));
    }

    protected void showEmailClientTemplate(@NonNull final String message) {
        showEmailClients(getString(R.string.app_support_email), getString(R.string.about_email_subject),
                message + UiUtils.getDeviceInfoString(null,false));
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/activities/base/BaseAppActivity.kt
package app.editors.manager.ui.activities.base

import android.os.Bundle
import android.os.Handler
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import app.editors.manager.R
import app.editors.manager.ui.interfaces.ContextDialogInterface
import lib.toolkit.base.managers.utils.ActivitiesUtils.showBrowser
import lib.toolkit.base.managers.utils.ActivitiesUtils.showEmail
import lib.toolkit.base.managers.utils.FragmentUtils.showFragment
import lib.toolkit.base.managers.utils.FragmentUtils.showSingleFragment
import lib.toolkit.base.managers.utils.UiUtils.getDeviceInfoString
import lib.toolkit.base.ui.activities.base.BaseActivity

abstract class BaseAppActivity : BaseActivity(), FragmentManager.OnBackStackChangedListener,
    ContextDialogInterface {

    private val TAG = javaClass.simpleName
    private var handler: Handler? = null
    private var finishRunnable: Runnable? = null
    private var isFinish = false
    private var isBackStackNotice = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        init()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(TAG_FINISH, isFinish)
        super.onSaveInstanceState(outState)
    }

    override fun onContextDialogOpen() {}
    override fun onDestroy() {
        super.onDestroy()
        finishRunnable?.let { runnable ->
            handler?.removeCallbacks(runnable)
        }
        supportFragmentManager.removeOnBackStackChangedListener(this)
    }

    override fun onBackStackChanged() {}

    private fun init() {
        initHandlers()
        initViews()
    }

    private fun initHandlers() {
        isBackStackNotice = false
        handler = Handler()
        finishRunnable = Runnable { isFinish = false }
    }

    private fun startResetFinishTimer() {
        finishRunnable?.let { runnable ->
            handler?.postDelayed(runnable, TIMER_FINISH.toLong())
        }
    }

    private fun initViews() {
        supportFragmentManager.addOnBackStackChangedListener(this)
        isFinish = false
    }

    private fun showEmailClients(to: String?, subject: String?, body: String?) {
        showEmail(this, getString(lib.toolkit.base.R.string.chooser_email_client), to!!, subject!!, body!!)
    }

    /*
     * Fragment operations
     * */
    protected fun showFragment(fragment: Fragment, tag: String?) {
        showFragment(supportFragmentManager, fragment, R.id.frame_container, tag, false)
    }

    protected fun showSingleFragment(fragment: Fragment, backStackTag: String?) {
        showSingleFragment(supportFragmentManager, fragment, R.id.frame_container, backStackTag!!)
    }

    /*
     * Helper methods
     * */
    protected fun showUrlInBrowser(url: String?) {
        showBrowser(this, getString(R.string.chooser_web_browser), url!!)
    }

    protected fun showEmailClientTemplate() {
        showEmailClients(
            getString(R.string.app_support_email), getString(lib.toolkit.base.R.string.about_email_subject),
            getDeviceInfoString(null, true)
        )
    }

    protected fun showEmailClientTemplate(message: String) {
        showEmailClients(
            getString(R.string.app_support_email), getString(lib.toolkit.base.R.string.about_email_subject),
            message + getDeviceInfoString(null, false)
        )
    }

    companion object {
        private const val TAG_FINISH = "TAG_FINISH"
        private const val TIMER_FINISH = 5000
    }
}